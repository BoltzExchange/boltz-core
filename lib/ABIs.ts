import ERC20ABI from '../out/ERC20.sol/ERC20.json';
import ERC20SwapABI from '../out/ERC20Swap.sol/ERC20Swap.json';
import EtherSwapABI from '../out/EtherSwap.sol/EtherSwap.json';
import RouterABI from '../out/Router.sol/Router.json';

export const ContractABIs = {
  ERC20: ERC20ABI.abi,
  ERC20Swap: ERC20SwapABI.abi,
  EtherSwap: EtherSwapABI.abi,
  Router: RouterABI.abi,
} as const;
