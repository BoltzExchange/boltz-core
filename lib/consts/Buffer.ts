export const Nonce = Buffer.from('00', 'hex');
export const EmptyScript = Buffer.alloc(0);
export const PrefixUnconfidential = Buffer.from('01', 'hex');
