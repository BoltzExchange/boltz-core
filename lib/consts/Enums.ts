export enum OutputType {
  Bech32 = 0,
  Compatibility = 1,
  Legacy = 2,
  Taproot = 3,
}
