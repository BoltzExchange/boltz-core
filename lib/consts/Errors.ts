const Errors = {
  TAPROOT_NOT_SUPPORTED: 'taproot not supported',
};

export default Errors;
