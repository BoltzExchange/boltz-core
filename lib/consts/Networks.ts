import { networks } from 'bitcoinjs-lib';

const bitcoinPrefix = '\\x18Bitcoin Signed Message:\n';
const litecoinPrefix = '\\x19Litecoin Signed Message:\n';
const dogecoinPrefix = '\\x19Dogecoin Signed Message:\n';

const testnetBip32 = {
  public: 0x019DA462,
  private: 0x019D9CFE,
};

const simnetBip32 = {
  public: 0x420BD3A,
  private: 0x488B21E,
};

const Networks = {
  // Bitcoin networks
  bitcoinMainnet: networks.bitcoin,
  bitcoinTestnet: networks.testnet,
  bitcoinSimnet: {
    messagePrefix: bitcoinPrefix,
    bip32: simnetBip32,
    bech32: 'sb',
    scriptHash: 0x7b,
    pubKeyHash: 0x3f,
    wif: 0x64,
  },
  bitcoinRegtest: networks.regtest,

  // Litecoin networks
  litecoinMainnet: {
    messagePrefix: litecoinPrefix,
    bip32: {
      private: 0x488ADE4,
      public: 0x488B21E,
    },
    bech32: 'ltc',
    scriptHash: 0x32,
    pubKeyHash: 0x30,
    wif: 0xb0,
  },
  litecoinTestnet: {
    messagePrefix: litecoinPrefix,
    bip32: testnetBip32,
    bech32: 'tltc',
    scriptHash: 0x3a,
    pubKeyHash: 0x6f,
    wif: 0xef,
  },
  litecoinSimnet: {
    messagePrefix: bitcoinPrefix,
    bip32: simnetBip32,
    bech32: 'sltc',
    scriptHash: 0x7b,
    pubKeyHash: 0x3f,
    wif: 0x64,
  },
  litecoinRegtest: {
    messagePrefix: litecoinPrefix,
    bip32: testnetBip32,
    bech32: 'rltc',
    scriptHash: 0x3a,
    pubKeyHash: 0x6f,
    wif: 0xef,
  },

  // Dogecoin networks
  dogecoinMainnet: {
    messagePrefix: dogecoinPrefix,
    bip32: {
      public: 0x02facafd,
      private: 0x02fac398,
    },
    pubKeyHash: 0x1e,
    scriptHash: 0x16,
    wif: 0x9e,
  },
  dogecoinTestnet: {
    messagePrefix: dogecoinPrefix,
    bip32: testnetBip32,
    pubKeyHash: 0x71,
    scriptHash: 0xc4,
    wif: 0xf1,
  },
  dogecoinRegtest: {
    messagePrefix: dogecoinPrefix,
    bip32: testnetBip32,
    pubKeyHash: 0x6f,
    scriptHash: 0xc4,
    wif: 0xef,
  },
};

export default Networks;
