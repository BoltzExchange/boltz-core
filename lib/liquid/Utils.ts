import { confidential, TxOutput } from 'liquidjs-lib';
import { confidentialLiquid } from './Init';

export const getOutputValue = (
  output: TxOutput & {
    blindingPrivKey?: Buffer;
  },
): number => {
  return output.blindingPrivKey
    ? Number(
        confidentialLiquid.unblindOutputWithKey(output, output.blindingPrivKey)
          .value,
      )
    : confidential.confidentialValueToSatoshi(output.value);
};
