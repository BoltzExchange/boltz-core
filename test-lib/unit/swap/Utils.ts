import Errors from '../../../lib/consts/Errors';
import { OutputType } from '../../../lib/consts/Enums';
import { p2wshOutput, p2shOutput, p2shP2wshOutput } from '../../../lib/swap/Scripts';

export const getScriptHashFunction = (type: OutputType): (scriptHex: Buffer) => Buffer => {
  switch (type) {
    case OutputType.Bech32: return p2wshOutput;
    case OutputType.Legacy: return p2shOutput;
    case OutputType.Compatibility: return p2shP2wshOutput;
    case OutputType.Taproot: throw Errors.TAPROOT_NOT_SUPPORTED;
  }
};
