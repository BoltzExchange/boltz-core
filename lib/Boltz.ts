import { targetFee } from './TargetFee';
import { OutputType } from './consts/Enums';
import Networks from './consts/Networks';
import * as Types from './consts/Types';
import {
  ClaimDetails,
  RefundDetails,
  ScriptElement,
  TransactionOutput,
} from './consts/Types';
import { init } from './init';
import Musig from './musig/Musig';
import { constructClaimTransaction } from './swap/Claim';
import { detectPreimage } from './swap/PreimageDetector';
import { constructRefundTransaction } from './swap/Refund';
import reverseSwapScript from './swap/ReverseSwapScript';
import reverseSwapTree, {
  extractClaimPublicKeyFromReverseSwapTree,
  extractRefundPublicKeyFromReverseSwapTree,
} from './swap/ReverseSwapTree';
import * as Scripts from './swap/Scripts';
import { detectSwap } from './swap/SwapDetector';
import swapScript from './swap/SwapScript';
import swapTree, {
  extractClaimPublicKeyFromSwapTree,
  extractRefundPublicKeyFromSwapTree,
} from './swap/SwapTree';
import { compareTrees } from './swap/SwapTreeCompare';
import * as SwapTreeSerializer from './swap/SwapTreeSerializer';
import * as SwapUtils from './swap/SwapUtils';
import * as TaprootUtils from './swap/TaprootUtils';

export {
  Musig,
  Types,
  Networks,
  OutputType,
  ClaimDetails,
  ScriptElement,
  RefundDetails,
  TransactionOutput,
  Scripts,
  SwapUtils,
  TaprootUtils,
  SwapTreeSerializer,
  init,
  swapTree,
  targetFee,
  swapScript,
  detectSwap,
  compareTrees,
  detectPreimage,
  reverseSwapTree,
  reverseSwapScript,
  constructClaimTransaction,
  constructRefundTransaction,
  extractClaimPublicKeyFromSwapTree,
  extractRefundPublicKeyFromSwapTree,
  extractClaimPublicKeyFromReverseSwapTree,
  extractRefundPublicKeyFromReverseSwapTree,
};
