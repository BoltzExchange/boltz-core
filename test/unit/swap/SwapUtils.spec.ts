import ops from '@boltz/bitcoin-ops';
import { crypto, Transaction } from 'bitcoinjs-lib';
import { OutputType } from '../../../lib/consts/Enums';
import { getHexBuffer, getHexString } from '../../../lib/Utils';
import {
  p2pkhOutput,
  p2shOutput,
  p2shP2wpkhOutput,
  p2shP2wshOutput,
  p2trOutput,
  p2wpkhOutput,
  p2wshOutput,
} from '../../../lib/swap/Scripts';
import {
  encodeCltv,
  encodeSignature,
  getOutputScriptType,
  scriptBuffersToScript,
  toPushdataScript,
} from '../../../lib/swap/SwapUtils';

describe('SwapUtils', () => {
  const publicKeyHash = getHexBuffer(
    '0000000000000000000000000000000000000000',
  );

  test.each`
    signature                                                                                                                             | flag
    ${'4e45e16932b8af514961a1d3a1a25fdf3f4f7732e9d624c6c61548ab5fb8cd41181522ec8eca07de4860a4acdd12909d831cc56cbbac4622082221a8768d1d09'} | ${Transaction.SIGHASH_ALL}
    ${'82235e21a2300022738dabb8e1bbd9d19cfb1e7ab8c30a23b0afbb8d178abcf324bf68e256c534ddfaf966bf908deb944305596f7bdcc38d69acad7f9c868724'} | ${Transaction.SIGHASH_ALL}
    ${'1cadddc2838598fee7dc35a12b340c6bde8b389f7bfd19a1252a17c4b5ed2d71c1a251bbecb14b058a8bd77f65de87e51c47e95904f4c0e9d52eddc21c1415ac'} | ${Transaction.SIGHASH_ALL}
    ${'1b19519b38ca1e6813cd25649ad36be8bc6a6f2ad9758089c429acd9ce0b572f3bf32193c8a3a3de1f847cd6e6eebf43c96df1ffa4d7fe920f8f71708920c65f'} | ${Transaction.SIGHASH_ALL}
  `('should encode signature $signature', ({ signature, flag }) => {
    const result = encodeSignature(flag, getHexBuffer(signature));
    expect(getHexString(result)).toMatchSnapshot();
  });

  test('should get formed scripts', () => {
    const result = scriptBuffersToScript([publicKeyHash, ops.OP_PUSHDATA1]);
    expect(getHexString(result)).toMatchSnapshot();
  });

  test('should get formed pushdata scripts', () => {
    const result = toPushdataScript([
      ops.OP_HASH160,
      crypto.hash160(publicKeyHash),
      ops.OP_EQUALVERIFY,
    ]);
    expect(getHexString(result)).toMatchSnapshot();
  });

  test.each`
    output                             | expected                                     | name
    ${p2wpkhOutput(publicKeyHash)}     | ${{ type: OutputType.Bech32, isSh: false }}  | ${'P2WKH'}
    ${p2pkhOutput(publicKeyHash)}      | ${{ type: OutputType.Legacy, isSh: false }}  | ${'P2PKH'}
    ${p2wshOutput(publicKeyHash)}      | ${{ type: OutputType.Bech32, isSh: true }}   | ${'P2WSH'}
    ${p2shOutput(publicKeyHash)}       | ${{ type: OutputType.Legacy, isSh: true }}   | ${'P2SH'}
    ${p2shP2wpkhOutput(publicKeyHash)} | ${{ type: OutputType.Legacy, isSh: true }}   | ${'P2SH nested P2WPKH'}
    ${p2shP2wshOutput(publicKeyHash)}  | ${{ type: OutputType.Legacy, isSh: true }}   | ${'P2SH nested P2WSH'}
    ${p2trOutput(publicKeyHash)}       | ${{ type: OutputType.Taproot, isSh: false }} | ${'P2TR'}
    ${[ops.OP_INVALIDOPCODE]}          | ${undefined}                                 | ${'OP_INVALIDOPCODE'}
    ${[ops.OP_RETURN]}                 | ${undefined}                                 | ${'OP_RETURN'}
  `(
    'should get the correct output type of $name output scripts',
    ({ output, expected }) => {
      expect(getOutputScriptType(output.outputScript || output)).toEqual(
        expected,
      );
    },
  );

  test.each([608861, 123, 321, 2345234, 790926])(
    'should encode CLTV @ block number %p',
    (blockHeight: number) => {
      expect(getHexString(encodeCltv(blockHeight))).toMatchSnapshot();
    },
  );
});
