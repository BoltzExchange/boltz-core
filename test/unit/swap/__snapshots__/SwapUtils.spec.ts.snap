// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SwapUtils should encode CLTV @ block number 123 1`] = `"7b"`;

exports[`SwapUtils should encode CLTV @ block number 321 1`] = `"4101"`;

exports[`SwapUtils should encode CLTV @ block number 608861 1`] = `"5d4a09"`;

exports[`SwapUtils should encode CLTV @ block number 790926 1`] = `"8e110c"`;

exports[`SwapUtils should encode CLTV @ block number 2345234 1`] = `"12c923"`;

exports[`SwapUtils should encode signature 1b19519b38ca1e6813cd25649ad36be8bc6a6f2ad9758089c429acd9ce0b572f3bf32193c8a3a3de1f847cd6e6eebf43c96df1ffa4d7fe920f8f71708920c65f 1`] = `"304402201b19519b38ca1e6813cd25649ad36be8bc6a6f2ad9758089c429acd9ce0b572f02203bf32193c8a3a3de1f847cd6e6eebf43c96df1ffa4d7fe920f8f71708920c65f01"`;

exports[`SwapUtils should encode signature 1cadddc2838598fee7dc35a12b340c6bde8b389f7bfd19a1252a17c4b5ed2d71c1a251bbecb14b058a8bd77f65de87e51c47e95904f4c0e9d52eddc21c1415ac 1`] = `"304502201cadddc2838598fee7dc35a12b340c6bde8b389f7bfd19a1252a17c4b5ed2d71022100c1a251bbecb14b058a8bd77f65de87e51c47e95904f4c0e9d52eddc21c1415ac01"`;

exports[`SwapUtils should encode signature 4e45e16932b8af514961a1d3a1a25fdf3f4f7732e9d624c6c61548ab5fb8cd41181522ec8eca07de4860a4acdd12909d831cc56cbbac4622082221a8768d1d09 1`] = `"304402204e45e16932b8af514961a1d3a1a25fdf3f4f7732e9d624c6c61548ab5fb8cd410220181522ec8eca07de4860a4acdd12909d831cc56cbbac4622082221a8768d1d0901"`;

exports[`SwapUtils should encode signature 82235e21a2300022738dabb8e1bbd9d19cfb1e7ab8c30a23b0afbb8d178abcf324bf68e256c534ddfaf966bf908deb944305596f7bdcc38d69acad7f9c868724 1`] = `"304502210082235e21a2300022738dabb8e1bbd9d19cfb1e7ab8c30a23b0afbb8d178abcf3022024bf68e256c534ddfaf966bf908deb944305596f7bdcc38d69acad7f9c86872401"`;

exports[`SwapUtils should get formed pushdata scripts 1`] = `"a914944f997c5553a6f3e1028e707c71b5fa0dd3afa788"`;

exports[`SwapUtils should get formed scripts 1`] = `"1400000000000000000000000000000000000000004c"`;
